impl/src/main/java/org/pentaho/platform/genericfile/DefaultGenericFileService.java

```diff
@@ -235,4 +237,24 @@ public List<IGenericFile> getDeletedFiles() throws OperationFailedException {
 
     return deletedFiles;
   }
+
+  @Override
+  public void deleteFilesPermanently( @NonNull List<GenericFilePath> paths ) throws OperationFailedException {
+    BatchDeleteOperationFailedException batchException =
+      new BatchDeleteOperationFailedException( "Error(s) occurred during deletion." );
+
+    for ( GenericFilePath path : paths ) {
+      try {
+        getOwnerFileProvider( path )
+          .orElseThrow( () -> new NotFoundException( String.format( "Path not found '%s'.", path ) ) )
+          .deleteFilePermanently( path );
+      } catch ( OperationFailedException e ) {
+        batchException.addFailedPath( path, e );
```

Comment:
I would prefer creating the `BatchDeleteOperationFailedException` exception here, lazily. Saves memory, but more importantly, the exception stack will capture the exception block location instead.